### ISSUE TITLE
How to determine an arbitrary orchestration/activity has reach its retry limits (when there is one)

### ISSUE THREAD

**Issue by jundayin** [Azure/durabletask#685]
When there is a retry options set for a scheduled orchestration/activity, the worker would retry the job until it exceeds the limit. 

In our code, we log and send ApplicationInsight exception request when an exception is caught inside the orchestration/activity. 

Ideally dev should receive alert when an exception thrown on the final run of the orchestration/activity but in practice when retry takes place, it become quite noisy because each retry would raise an alert. 

I wonder is there a way from client or worker to tell whether the current orchestration/activity is on its last run as in dev only receive one alert for one orchestration instance. 

(I understand we can tell whether it is re-run by the orchestration instance Id, but it is quite hacky and it is the last approach I would like to pursue)

**Comment by jundayin** [comment]
@cgillum Sorry for tagging people directly, but we would like to hear from the owner before we do something from our end.

**Comment by cgillum** [comment]
I don't believe the current retry count is surfaced to the application code anywhere, so it's not something I think you can achieve directly. However, if a retry limit is exceeded then the exception thrown by the activity or orchestrator will propagate to the parent orchestration. You could then catch the exception in the parent orchestration and send the notification indicating that something failed after maxing-out all the retry attempts. Would this approach work for you?

**Comment by jundayin** [comment]
This is one option, but doesn't work well when orchestration and activity are reused by different combination upon business logic.

Let's say a business flow Alpha: Orchestration A -> Sub-Orchestration B -> Activity x. In this case, A is the parent orchestration and it is responsible of sending failure notification and B or x should not take care of it. That makes sense.

But in another business flow Beta: Orchestration B -> Activity y.  Here B is the parent orchestration and it should be the one which sends the notification. Now if you look back to flow Alpha, A and B both sends notification and noise generated.

I'm thinking of adding one more property from `OrchestrationInstance` for the flag indicating if it is the last run. Would it be a breaking change?

**Comment by cgillum** [comment]
Adding status information to `OrchestrationInstance` is not recommended because this data type is expected to be immutable: it's persisted in the history.

It sounds like you want the top-level orchestration to be the one to send a notification if there is a failure? A simpler approach might be to expose a `ParentOrchestrationInstance` property to the `OrchestrationContext` class which could be used to determine whether the current orchestration has a parent or not. If it doesn't (i.e. this new property is `null`) then you could send a notification if there is a failure.
